src/catchup/CatchupManagerImpl.cpp

```diff
@@ -439,8 +499,9 @@ void
 CatchupManagerImpl::tryApplySyncingLedgers()
 {
     ZoneScoped;
-    auto const& ledgerHeader =
-        mApp.getLedgerManager().getLastClosedLedgerHeader();
+    releaseAssert(threadIsMain());
+    uint32_t nextToClose = *mLastQueuedToApply + 1;
```

Comment:
I don't know that this is right? Or rather the condition below that checks it isn't right? mLastQueuedToApply is basically "the back of the queue", right? And we're taking lcd off the _front_ of the queue?

Let me do an example: if the queue is, say, {4, 5, 6}, then LCL is 3, lcd is 4, mLastQueuedToApply is 6, and so nextToClose is 7. The condition we check below (nextToClose != lcd.getLegerSeq()) will be comparing 7 with 4, they are unequal, so we break.

In the old code (before this change) it was comparing LCL+1 with lcd, which will be comparing 4 with 4, which are equal, so we break. I think the old code is correct.

But then .. I am a little surprised if the new code works at all. Even when the queue is just a single entry, say {4}, then LCL is 3, lcd is 4, mLastQueuedToApply is also 4, nextToClose is 5, and so again it will compare 5 with 4, they are unequal, it will break.

What am I misunderstanding here?

Reply:
So mLastQueuedToApply tracks the back of the queue of ledgers currently being applied in the background. It's different from the queue tracked by CM (sorry, many queues here!). CM queue is "any ledgers we received from the network", which may arrive out of order, and the queue may contain gaps. mLastQueuedToApply tracks the queue of strictly sequential ledgers that are guaranteed to be applied in the background (lambdas posted to ledgerCloseThread via asio). So once we've determined that a buffered ledger _can_ be applied, it's popped from buffered ledgers queue, and scheduled to be applied in the background (and mLastQueuedToApply is incremented)